src/reconcile/rec_write.c

```diff
@@ -5824,32 +5826,50 @@ __rec_row_leaf_insert(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins)
 	val = &r->v;
 
 	for (; ins != NULL; ins = WT_SKIP_NEXT(ins)) {
-		/* Look for an update, if nothing is visible, we're done. */
 		WT_RET(__rec_txn_read(session, r, ins, NULL, NULL, &upd));
-		if (upd == NULL)
-			continue;
+		if (upd == NULL) {
+			/*
+			 * Look for an update. If nothing is visible and not in
+			 * evict/restore, there's no work to do.
+			 */
+			if (!F_ISSET(r, WT_EVICT_UPDATE_RESTORE))
+				continue;
 
-		switch (upd->type) {
-		case WT_UPDATE_DELETED:
-			continue;
-		case WT_UPDATE_MODIFIED:
-			/* Impossible slot, there's no backing on-page item. */
-			cbt->slot = UINT32_MAX;
-			WT_RET(__wt_value_return(session, cbt, upd));
-			WT_RET(__rec_cell_build_val(session, r,
-			    cbt->iface.value.data,
-			    cbt->iface.value.size, (uint64_t)0));
-			break;
-		case WT_UPDATE_STANDARD:
-			if (upd->size == 0)
-				val->len = 0;
-			else
+			/*
+			 * When doing evict/restore, move the insert key to the
+			 * page, with an empty value (this allows us to split
+			 * the page if there's a huge, pinned insert list). The
+			 * on-page key must never be read, make sure there is a
```

Comment:
I believe this is the problem. What rollback_to_stable is doing, is moving the visibility point backward in time, so it will mark the UPDATE as WT_TXN_ABORTED. Though.. I'm surprised I see the particular symptoms I'm seeing, which are that the page has no WT_UPDATE structures, not WT_UPDATE structures with WT_TXN_ABORT. So there is something else happening here as well, I'm suspicious that there are two update-restores happening, the second one is happening on a page that only contained WT_TXN_ABORTED updates, and no valid on-disk entries - I didn't get far enough in to confirm that, but I used the following code changes that helped my understanding:

```
diff --git a/src/btree/bt_curnext.c b/src/btree/bt_curnext.c
index eb8a258..080c4f3 100644
--- a/src/btree/bt_curnext.c
+++ b/src/btree/bt_curnext.c
@@ -298,6 +298,7 @@ __cursor_row_next(WT_CURSOR_BTREE *cbt, bool newpage)
                goto new_insert;
        }

+       WT_RET(__wt_debug_tree_all(session, (WT_BTREE*)session->dhandle->handle, NULL, "tmp2.txt"));
        /* Move to the next entry and return the item. */
        for (;;) {
                /*
diff --git a/src/btree/bt_debug.c b/src/btree/bt_debug.c
index d91ac02..72d651e 100644
--- a/src/btree/bt_debug.c
+++ b/src/btree/bt_debug.c
@@ -207,7 +207,7 @@ __debug_config(WT_SESSION_IMPL *session, WT_DBG *ds, const char *ofile)
                WT_RET(__wt_scr_alloc(session, 512, &ds->msg));
                ds->f = __dmsg_event;
        } else {
-               if ((ds->fp = fopen(ofile, "w")) == NULL)
+               if ((ds->fp = fopen(ofile, "a")) == NULL)
                        return (EIO);
                __wt_stream_set_line_buffer(ds->fp);
                ds->f = __dmsg_file;
diff --git a/src/evict/evict_page.c b/src/evict/evict_page.c
index 03d6e9a..402e2fe 100644
--- a/src/evict/evict_page.c
+++ b/src/evict/evict_page.c
@@ -583,6 +583,11 @@ __evict_review(
        *flagsp = flags;
        WT_RET(ret);

+       if (LF_ISSET(WT_EVICT_UPDATE_RESTORE)) {
+               fprintf(stderr, "Did an update restore eviction on %p\n", page);
+               WT_RET(__wt_debug_tree_all(session, (WT_BTREE*)session->dhandle->handle, NULL, "tmp.txt"));
+       }
+
        /*
         * Success: assert the page is clean or reconciliation was configured
         * for update/restore. If the page is clean, assert that reconciliation
diff --git a/src/txn/txn_rollback_to_stable.c b/src/txn/txn_rollback_to_stable.c
index e19bbc7..5b8f71f 100644
--- a/src/txn/txn_rollback_to_stable.c
+++ b/src/txn/txn_rollback_to_stable.c
@@ -415,6 +415,7 @@ __wt_txn_rollback_to_stable(WT_SESSION_IMPL *session, const char *cfg[])
        conn = S2C(session);
        WT_RET(__txn_rollback_to_stable_check(session));

+       fprintf(stderr, "Rollback to stable starting\n");
        /* Allocate a non-durable btree bitstring */
        WT_RET(__bit_alloc(session,
            conn->next_file_id, &conn->stable_rollback_bitstring));
@@ -428,6 +429,7 @@ __wt_txn_rollback_to_stable(WT_SESSION_IMPL *session, const char *cfg[])
         * lookaside records should be removed.
         */
        WT_ERR(__txn_rollback_to_stable_lookaside_fixup(session));
+       fprintf(stderr, "Rollback to stable finished\n");
 err:   __wt_free(session, conn->stable_rollback_bitstring);
        return (ret);
 #endif
diff --git a/test/suite/test_timestamp04.py b/test/suite/test_timestamp04.py
index 3af0fee..32f3deb 100644
--- a/test/suite/test_timestamp04.py
+++ b/test/suite/test_timestamp04.py
@@ -64,6 +64,9 @@ class test_timestamp04(wttest.WiredTigerTestCase, suite_subprocess):
             session.begin_transaction(txn_config)
         c = session.open_cursor(self.uri, None)
         if missing == False:
+            while c.next() == 0:
+                print 'K: ' + str(c.get_key()) + ' V: ' + str(c.get_value())
+            c.reset()
             actual = dict((k, v) for k, v in c if v != 0)
             #print expected
             #print actual
@@ -95,7 +98,7 @@ class test_timestamp04(wttest.WiredTigerTestCase, suite_subprocess):
         c = self.session.open_cursor(self.uri)

         # Insert keys each with timestamp=key, in some order
-        key_range = 10000
+        key_range = 1000
         keys = range(1, key_range + 1)

         for k in keys:
@@ -112,10 +115,10 @@ class test_timestamp04(wttest.WiredTigerTestCase, suite_subprocess):

         # Now check that we see the expected state when reading at each
         # timestamp
-        self.check(self.session, 'read_timestamp=' + timestamp_str(key_range / 2),
-            dict((k, 1) for k in keys[:(key_range / 2)]))
-        self.check(self.session, 'read_timestamp=' + timestamp_str(key_range / 2),
-            dict((k, 1) for k in keys[(key_range / 2 + 1):]), missing=True)
+#        self.check(self.session, 'read_timestamp=' + timestamp_str(key_range / 2),
+#            dict((k, 1) for k in keys[:(key_range / 2)]))
+#        self.check(self.session, 'read_timestamp=' + timestamp_str(key_range / 2),
+#            dict((k, 1) for k in keys[(key_range / 2 + 1):]), missing=True)

         # Bump the oldest timestamp, we're not going back...
         self.conn.set_timestamp('oldest_timestamp=' + timestamp_str(key_range / 2))
@@ -127,8 +130,8 @@ class test_timestamp04(wttest.WiredTigerTestCase, suite_subprocess):
             self.session.commit_transaction('commit_timestamp=' + timestamp_str(k + key_range))

         # Now we should have: keys 1-100 with value 2
-        self.check(self.session, 'read_timestamp=' + timestamp_str(2 * key_range),
-            dict((k, 2) for k in keys[:]))
+#        self.check(self.session, 'read_timestamp=' + timestamp_str(2 * key_range),
+#            dict((k, 2) for k in keys[:]))

         # Rollback a quater of the new commits
         self.conn.set_timestamp('stable_timestamp=' + timestamp_str(1 + key_range + key_range / 4))
```

The output to console is:

```
test_timestamp04.test_timestamp04.test_basic -s 2 (row) ... 
Rollback to stable starting
Rollback to stable finished
Did an update restore eviction on 0x1dc8e80
Did an update restore eviction on 0x1dc3a00
Rollback to stable starting
Rollback to stable finished
Did an update restore eviction on 0x1f27b40
ERROR
```

There are also two files created with dumps of the pages from cache. `build_posix/WT_TEST/test_timestamp04.0/tmp.txt` contains the content in cache immediately after an eviction (I don't think this includes the page that has just been evicted) , and `build_posix/WT_TEST/test_timestamp04.0/tmp2.txt` contains the state of the pages when starting the cursor traversal.  That shows a page with records like:

```
K {#c1#b3}
        V {#81}
        K {#c1#b4}
        V {#81}
        K {#c2:}
        V {}
        K {#c2;}
        V {}
        K {#c2<}
        V {}
```

Which isn't expected in this workload - there shouldn't be 0 length values.

Reply:
@agorrod, if you have a few minutes, let's talk this over. What's happening here is that we're inserting 0-length values on the pages, but at the same time adding globally visible updates in front of those 0-length values, so no cursor retrieval should ever see them (although page-dump will display them of course).